### PR TITLE
Handle precompiled contracts errors

### PR DESCRIPTION
**Pull request by patogallaiovlabs**
## Description

This is a VM level issue fix.
In case of failure in precompiled calls from a contract, we are not pushing a Zero to the Stack and the exception bubbles up failing the execution of the caller too, spending all the gas of both executions.
What we are improving in this fix, is the way of handling those errors, not bubbling up the exception and pushing a Zero into the stack, consuming only the necessary gas for the precompiled call.

## Motivation and Context

We want to make sure we can handle the errors for precompiled calls in the best possible way. In the way it was implemented, there was no way of doing it, thats why we add this handler.

## How Has This Been Tested?

Unit testing, DSL tests and manual testing.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Requires Activation Code (Hard Fork)

* **Other information**:

### CODE REVIEW

File: rskj-core/src/main/java/org/ethereum/vm/program/Program.java [hunk 1]
```diff
@@ -1351,7 +1349,7 @@ public int verifyBeginSub(DataWord nextPC) {
         return ret;
     }
 
-    public void callToPrecompiledAddress(MessageCall msg, PrecompiledContract contract, ActivationConfig.ForBlock activations) {
+    public void callToPrecompiledAddress(MessageCall msg, PrecompiledContract contract) {
```
Comment (ajlopezrsk) [inline rskj-core/src/main/java/org/ethereum/vm/program/Program.java:1352]: No PrecompiledContractException throws? Why?

File: rskj-core/src/main/java/org/ethereum/vm/VM.java [hunk 2]
```diff
@@ -1490,9 +1490,9 @@ protected void doCALL(){
         PrecompiledContracts.PrecompiledContract precompiledContract = precompiledContracts.getContractForAddress(activations, codeAddress);
 
         if (precompiledContract != null) {
-            program.callToPrecompiledAddress(msg, precompiledContract, activations);
+            program.callToPrecompiledAddress(msg, precompiledContract);
```
Comment (ajlopezrsk) [inline rskj-core/src/main/java/org/ethereum/vm/VM.java:1493]: No treatment of PrecompiledContractException?

File: rskj-core/src/main/java/org/ethereum/config/blockchain/upgrades/ConsensusRule.java [hunk 3]
```diff
@@ -55,7 +55,9 @@
     RSKIP172("rskip172"),
     RSKIP174("rskip174"),
     RSKIP180("rskip180"),
-    RSKIPUMM("rskipUMM");
+    RSKIPUMM("rskipUMM"),
+    // TODO: change activation code RSKIPNEW.
+    RSKIPNEW("rskipNEW"); //Handle error in Precompile Contracts execution.
```
[the remark this answers is not in the document]
Comment (patogallaiovlabs) [reply rskj-core/src/main/java/org/ethereum/config/blockchain/upgrades/ConsensusRule.java:60]: We are on it :)

File: rskj-core/src/main/java/org/ethereum/vm/program/Program.java (hunk 1 above)
Comment (patogallaiovlabs) [reply rskj-core/src/main/java/org/ethereum/vm/program/Program.java:1352]: We wanted to keep consistency with the behaviour of "callToAddress()", which handles all the errors internally, it doesn't bubble up the error but push a Zero into the stack.

File: rskj-core/src/main/java/org/ethereum/config/blockchain/upgrades/ConsensusRule.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (fedejinich) [reply rskj-core/src/main/java/org/ethereum/config/blockchain/upgrades/ConsensusRule.java:60]: https://github.com/rsksmart/RSKIPs/pull/197

**Comment by fedejinich** [comment]
https://github.com/rsksmart/rskj/pull/1392/commits/0730e8728be82e25d34186e10621843938257c0f on this commit I've introduced [this](https://github.com/rsksmart/rskj/issues/1391) new feature

File: rskj-core/src/main/java/org/ethereum/config/blockchain/upgrades/ConsensusRule.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (fedejinich) [reply rskj-core/src/main/java/org/ethereum/config/blockchain/upgrades/ConsensusRule.java]: https://github.com/rsksmart/rskj/pull/1392/commits/11d5f369b2ae78b5afa3e73b0b63bc45bb0bd89a renamed

**Review by ajlopezrsk — DISMISSED** [review]

**Review by ajlopezrsk — DISMISSED** [review]

File: rskj-core/src/test/java/co/rsk/peg/performance/FederationChangeTest.java [hunk 4]
```diff
@@ -84,14 +87,15 @@ private void createFederation_winner(int times, ExecutionStats stats) {
     }
 
     @Test
-    public void addFederatorPublicKey() {
+    public void addFederatorPublicKey() throws VMException {
         ExecutionStats stats = new ExecutionStats("addFederatorPublicKey");
         addFederatorPublicKey_noWinner(200, stats);
         addFederatorPublicKey_winner(200, stats);
         BridgePerformanceTest.addStats(stats);
+        Assert.assertTrue(true);
```
Comment (donequis) [inline rskj-core/src/test/java/co/rsk/peg/performance/FederationChangeTest.java:95]: ?

File: rskj-core/src/test/java/co/rsk/peg/performance/FederationChangeTest.java [hunk 5]
```diff
@@ -119,14 +123,15 @@ private void addFederatorPublicKey_winner(int times, ExecutionStats stats) {
     }
 
     @Test
-    public void commitFederation() {
+    public void commitFederation() throws VMException {
         ExecutionStats stats = new ExecutionStats("commitFederation");
         commitFederation_noWinner(200, stats);
         commitFederation_winner(200, stats);
         BridgePerformanceTest.addStats(stats);
+        Assert.assertTrue(true);
```
Comment (donequis) [inline rskj-core/src/test/java/co/rsk/peg/performance/FederationChangeTest.java:131]: ?

File: rskj-core/src/test/java/co/rsk/peg/performance/FederationChangeTest.java [hunk 6]
```diff
@@ -151,14 +156,15 @@ private void commitFederation_winner(int times, ExecutionStats stats) {
     }
 
     @Test
-    public void rollbackFederation() {
+    public void rollbackFederation() throws VMException {
         ExecutionStats stats = new ExecutionStats("rollbackFederation");
         rollbackFederation_noWinner(200, stats);
         rollbackFederation_winner(200, stats);
         BridgePerformanceTest.addStats(stats);
+        Assert.assertTrue(true);
```
Comment (donequis) [inline rskj-core/src/test/java/co/rsk/peg/performance/FederationChangeTest.java:164]: ?

File: rskj-core/src/test/java/co/rsk/peg/performance/LockWhitelistTest.java [hunk 7]
```diff
@@ -45,25 +47,27 @@
     private static final ECKey authorizedWhitelistChanger = ECKey.fromPrivate(Hex.decode("3890187a3071327cee08467ba1b44ed4c13adb2da0d5ffcc0563c371fa88259c"));
 
     @Test
-    public void getLockWhitelistSize() throws IOException {
+    public void getLockWhitelistSize() throws IOException, VMException {
         ExecutionStats stats = new ExecutionStats("getLockWhitelistSize");
         executeTestCase((int executionIndex) -> Bridge.GET_LOCK_WHITELIST_SIZE.encode(), "getLockWhitelistSize", 200, stats);
         BridgePerformanceTest.addStats(stats);
+        Assert.assertTrue(true);
```
Comment (donequis) [inline rskj-core/src/test/java/co/rsk/peg/performance/LockWhitelistTest.java:54]: ?

**Comment by donequis** [comment]
It looks good. There are some Assert.assertTrue(true); in tests that I don't understand

File: rskj-core/src/test/java/co/rsk/peg/performance/LockWhitelistTest.java (hunk 7 above)
Comment (patogallaiovlabs) [reply rskj-core/src/test/java/co/rsk/peg/performance/LockWhitelistTest.java:54]: Was just for sonar not complaining... but I will change them to smthg with more sense.

**Review by ajlopezrsk — APPROVED** [review]

File: rskj-core/src/main/java/org/ethereum/vm/program/Program.java [hunk 8]
```diff
@@ -1424,28 +1425,73 @@ public void callToPrecompiledAddress(MessageCall msg, PrecompiledContract contra
...
-            } else if (!getActivations().isActive(ConsensusRule.RSKIP150)) {
-                this.memorySave(msg.getOutDataOffs().intValue(), out);
+    /**
+     * This is after RSKIP197, where we fix the way in which error is handled after a precompiled execution.
+     */
+    private void executePrecompiledAndHandleError(PrecompiledContract contract, MessageCall msg, long requiredGas, Repository track, byte[] data) {
+        try {
+            logger.trace("Executing Precompiled contract...");
+            byte[] out = contract.execute(data);
+            if (getActivations().isActive(ConsensusRule.RSKIP90)) {
```
[the remark this answers is not in the document]
Comment (patogallaiovlabs) [reply rskj-core/src/main/java/org/ethereum/vm/program/Program.java:1469]: Nice catch

[the remark this answers is not in the document]
Comment (donequis) [reply rskj-core/src/main/java/org/ethereum/vm/program/Program.java]: Theoretically you could have a regtest where it is false, but that doesn't make much sense really.

**Review by donequis — DISMISSED** [review]

**Review by josedahlquist — APPROVED** [review]
Good job!

File: rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java [hunk 9]
```diff
@@ -1992,9 +1993,14 @@ public boolean increaseLockingCap(Transaction tx, Coin newCap) {
...
-    public void registerBtcCoinbaseTransaction(byte[] btcTxSerialized, Sha256Hash blockHash, byte[] pmtSerialized, Sha256Hash witnessMerkleRoot, byte[] witnessReservedValue) throws IOException, BlockStoreException {
+    public void registerBtcCoinbaseTransaction(byte[] btcTxSerialized, Sha256Hash blockHash, byte[] pmtSerialized, Sha256Hash witnessMerkleRoot, byte[] witnessReservedValue) throws VMException {
         Context.propagate(btcContext);
-        this.ensureBtcBlockStore();
+        try{
+            this.ensureBtcBlockStore();
+        }catch (BlockStoreException | IOException e) {
+            logger.warn("Exception in registerBtcCoinbaseTransaction", e);
+            throw new VMException("Exception in registerBtcCoinbaseTransaction", e);
+        }
```
[the remark this answers is not in the document]
Comment (patogallaiovlabs) [reply rskj-core/src/main/java/co/rsk/peg/BridgeSupport.java:2003]: Not sure if catching inside, could break the behaviour of other methods like "getBtcTransactionConfirmationsGetCost". So, I thought that the safest way was this one, apart from out of scope. Let me know your thoughts.

**Review by Vovchyk — APPROVED** [review]

**Pull request merged**